File: sdk/experimental/php/src/Output/Registry.php
<?php

declare(strict_types=1);

namespace HopTop\Kit\Output;

use HopTop\Kit\Output\Formatter\Formatter;
use InvalidArgumentException;

/**
 * Holds Formatter implementations keyed by Formatter::key().
 *
 * register() throws on duplicate key (PHP idiom for "this argument can't
 * be accepted in current state"); adopters intentionally replacing a
 * built-in must call override().
 *
 * Mirrors py Registry, ts Registry, go *Registry. Includes a static
 * default() accessor that returns the process-wide singleton — built-in
 * formatters register against it at bootstrap time.
 */
final class Registry
{
    /** @var array<string,Formatter> */
    private array $byKey = [];

    private static ?Registry $default = null;

    public static function default(): Registry
    {
        if (self::$default === null) {
            self::$default = new Registry();
            Builtins::register(self::$default);
        }
        return self::$default;
    }

    /**
     * Replace the process-wide default registry — tests use this to start
     * from a clean slate then restore. Returns the previous default so
     * callers can restore without leaking state.
     */
    public static function setDefault(Registry $r): Registry
    {
        $prev = self::$default ?? new Registry();
        self::$default = $r;
        return $prev;
    }

    public function register(Formatter $f): void
    {
        $this->validate($f);
        $key = $f->key();
        if (isset($this->byKey[$key])) {
            throw new InvalidArgumentException(sprintf(
                "output: formatter '%s' already registered (use override to replace)",
                $key,
            ));
        }
        $this->byKey[$key] = $f;
    }

    public function override(Formatter $f): void
    {
        $this->validate($f);
        $this->byKey[$f->key()] = $f;
    }

    public function lookup(string $key): ?Formatter
    {
        return $this->byKey[$key] ?? null;
    }

    /**
     * @return list<string> registered keys, sorted for stable output
     */
    public function keys(): array
    {
        $keys = array_keys($this->byKey);
        sort($keys);
        return array_values($keys);
    }

    /**
     * @return list<Formatter> formatters in key order
     */
    public function formatters(): array
    {
        $keys = $this->keys();
        return array_values(array_map(fn (string $k) => $this->byKey[$k], $keys));
    }

    /**
     * @return array<string,string> ext (lowercase, no leading dot) → formatter key
     */
    public function extensionMap(): array
    {
        $out = [];
        foreach ($this->keys() as $k) {
            foreach ($this->byKey[$k]->extensions() as $ext) {
                $out[strtolower(ltrim($ext, '.'))] = $k;
            }
        }
        return $out;
    }

    private function validate(Formatter $f): void
    {
        if ($f->key() === '') {
            throw new InvalidArgumentException('output: formatter key is empty');
        }
    }
}

File: sdk/experimental/rs/src/output/registry.rs
//! Registry — holds Formatter implementations keyed by [`Formatter::key`].
//!
//! Mirrors py `Registry`, ts `Registry`, go `*Registry`, php `Registry`.

use std::collections::BTreeMap;
use std::sync::{Arc, OnceLock, RwLock};

use thiserror::Error;

use super::formatter::Formatter;

/// Errors returned by [`Registry::register`].
#[derive(Debug, Error)]
pub enum RegisterError {
    #[error("output: formatter key is empty")]
    EmptyKey,
    #[error("output: formatter '{0}' already registered (use override to replace)")]
    Duplicate(String),
}

/// Holds Formatter implementations keyed by [`Formatter::key`].
///
/// `register` returns an error on duplicate; adopters intentionally
/// replacing a built-in must call `override_with`.
pub struct Registry {
    inner: RwLock<BTreeMap<&'static str, Arc<dyn Formatter>>>,
}

impl Registry {
    pub fn new() -> Self {
        Self {
            inner: RwLock::new(BTreeMap::new()),
        }
    }

    pub fn register(&self, f: Arc<dyn Formatter>) -> Result<(), RegisterError> {
        let key = f.key();
        if key.is_empty() {
            return Err(RegisterError::EmptyKey);
        }
        let mut g = self.inner.write().expect("registry rwlock poisoned");
        if g.contains_key(key) {
            return Err(RegisterError::Duplicate(key.to_string()));
        }
        g.insert(key, f);
        Ok(())
    }

    /// Replace (or insert) the formatter under `f.key()`. Used by
    /// adopters intentionally swapping out a built-in.
    pub fn override_with(&self, f: Arc<dyn Formatter>) -> Result<(), RegisterError> {
        let key = f.key();
        if key.is_empty() {
            return Err(RegisterError::EmptyKey);
        }
        self.inner
            .write()
            .expect("registry rwlock poisoned")
            .insert(key, f);
        Ok(())
    }

    pub fn lookup(&self, key: &str) -> Option<Arc<dyn Formatter>> {
        self.inner
            .read()
            .expect("registry rwlock poisoned")
            .get(key)
            .cloned()
    }

    /// All registered keys, sorted for stable output.
    pub fn keys(&self) -> Vec<&'static str> {
        self.inner
            .read()
            .expect("registry rwlock poisoned")
            .keys()
            .copied()
            .collect()
    }

    /// All registered formatters in key order.
    pub fn formatters(&self) -> Vec<Arc<dyn Formatter>> {
        self.inner
            .read()
            .expect("registry rwlock poisoned")
            .values()
            .cloned()
            .collect()
    }

    /// `ext` (lowercase, no leading dot) → formatter key.
    pub fn extension_map(&self) -> BTreeMap<String, &'static str> {
        let mut out = BTreeMap::new();
        let g = self.inner.read().expect("registry rwlock poisoned");
        for (k, f) in g.iter() {
            for ext in f.extensions() {
                out.insert(ext.trim_start_matches('.').to_ascii_lowercase(), *k);
            }
        }
        out
    }
}

impl Default for Registry {
    fn default() -> Self {
        Self::new()
    }
}

/// Process-wide default registry. Built-ins (JSON, YAML) are registered
/// against it lazily on first access.
pub fn default_registry() -> Arc<Registry> {
    static DEFAULT: OnceLock<Arc<Registry>> = OnceLock::new();
    DEFAULT
        .get_or_init(|| {
            let r = Arc::new(Registry::new());
            super::builtins::register_all(&r);
            r
        })
        .clone()
}
